File: src/lib/db/sessions.ts
// Session management functions
import 'server-only';
import { getDatabase } from './connection';
import { randomBytes } from 'crypto';

export interface Session {
  token: string;
  created_at: string;
  expires_at: string;
  last_used_at: string;
}

const ONE_YEAR_IN_MS = 365 * 24 * 60 * 60 * 1000;

/**
 * Create a new session token
 */
export function createSession(): Session {
  const db = getDatabase();
  const token = randomBytes(32).toString('base64url');
  const now = new Date();
  const expiresAt = new Date(now.getTime() + ONE_YEAR_IN_MS);

  const session: Session = {
    token,
    created_at: now.toISOString(),
    expires_at: expiresAt.toISOString(),
    last_used_at: now.toISOString(),
  };

  db.prepare(`
    INSERT INTO sessions (token, created_at, expires_at, last_used_at)
    VALUES (?, ?, ?, ?)
  `).run(session.token, session.created_at, session.expires_at, session.last_used_at);

  return session;
}

/**
 * Validate a session token
 * Returns the session if valid, null if invalid or expired
 */
export function validateSession(token: string): Session | null {
  const db = getDatabase();

  const session = db.prepare(`
    SELECT token, created_at, expires_at, last_used_at
    FROM sessions
    WHERE token = ?
  `).get(token) as Session | undefined;

  if (!session) {
    return null;
  }

  // Check if session is expired
  const now = new Date();
  const expiresAt = new Date(session.expires_at);

  if (now > expiresAt) {
    // Delete expired session
    deleteSession(token);
    return null;
  }

  // Update last used timestamp
  db.prepare(`
    UPDATE sessions
    SET last_used_at = ?
    WHERE token = ?
  `).run(now.toISOString(), token);

  return {
    ...session,
    last_used_at: now.toISOString(),
  };
}

/**
 * Delete a session
 */
export function deleteSession(token: string): void {
  const db = getDatabase();
  db.prepare('DELETE FROM sessions WHERE token = ?').run(token);
}

/**
 * Delete all expired sessions (cleanup)
 */
export function deleteExpiredSessions(): number {
  const db = getDatabase();
  const now = new Date().toISOString();

  const result = db.prepare(`
    DELETE FROM sessions
    WHERE expires_at < ?
  `).run(now);

  return result.changes;
}

/**
 * Verify password (from environment variable)
 */
export function verifyPassword(password: string): boolean {
  const correctPassword = process.env.AUTH_PASSWORD;

  if (!correctPassword) {
    throw new Error('AUTH_PASSWORD environment variable is not set');
  }

  return password === correctPassword;
}

File: backend/db/sessions.go
package db

import (
	"context"
	"database/sql"
	"time"
)

const (
	// SessionDuration is the default session lifetime (30 days)
	SessionDuration = 30 * 24 * time.Hour
)

// CreateSession creates a new session in the database
func (d *DB) CreateSession(ctx context.Context, id string) (*Session, error) {
	now := NowMs()
	expiresAt := time.Now().Add(SessionDuration).UnixMilli()

	if err := d.Write(ctx, func(tx *sql.Tx) error {
		_, err := tx.Exec(`
			INSERT INTO sessions (id, created_at, expires_at, last_used_at)
			VALUES (?, ?, ?, ?)
		`, id, now, expiresAt, now)
		return err
	}); err != nil {
		return nil, err
	}

	return &Session{
		ID:         id,
		CreatedAt:  now,
		ExpiresAt:  expiresAt,
		LastUsedAt: now,
	}, nil
}

// GetSession retrieves a session by ID, returns nil if not found or expired
func (d *DB) GetSession(id string) (*Session, error) {
	var s Session
	err := d.conn.QueryRow(`
		SELECT id, created_at, expires_at, last_used_at
		FROM sessions
		WHERE id = ? AND expires_at > ?
	`, id, NowMs()).Scan(&s.ID, &s.CreatedAt, &s.ExpiresAt, &s.LastUsedAt)

	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	return &s, nil
}

// TouchSession updates the last_used_at timestamp for a session
func (d *DB) TouchSession(ctx context.Context, id string) error {
	return d.Write(ctx, func(tx *sql.Tx) error {
		_, err := tx.Exec(`
			UPDATE sessions
			SET last_used_at = ?
			WHERE id = ?
		`, NowMs(), id)
		return err
	})
}

// DeleteSession removes a session from the database
func (d *DB) DeleteSession(ctx context.Context, id string) error {
	return d.Write(ctx, func(tx *sql.Tx) error {
		_, err := tx.Exec(`DELETE FROM sessions WHERE id = ?`, id)
		return err
	})
}

// DeleteExpiredSessions removes all expired sessions
func (d *DB) DeleteExpiredSessions(ctx context.Context) (int64, error) {
	var affected int64
	err := d.Write(ctx, func(tx *sql.Tx) error {
		result, err := tx.Exec(`DELETE FROM sessions WHERE expires_at <= ?`, NowMs())
		if err != nil {
			return err
		}
		affected, err = result.RowsAffected()
		return err
	})
	if err != nil {
		return 0, err
	}
	return affected, nil
}

// ExtendSession extends the expiration time of a session
func (d *DB) ExtendSession(ctx context.Context, id string) error {
	expiresAt := time.Now().Add(SessionDuration).UnixMilli()
	return d.Write(ctx, func(tx *sql.Tx) error {
		_, err := tx.Exec(`
			UPDATE sessions
			SET expires_at = ?, last_used_at = ?
			WHERE id = ?
		`, expiresAt, NowMs(), id)
		return err
	})
}
